src/sage/rings/function_field/drinfeld_modules/finite_drinfeld_module.py

```diff
@@ -391,39 +436,35 @@ def _frobenius_charpoly_gekeler(self, var='X'):
         for i in range(r-1):
             block_shifts.append(block_shifts[-1] + shifts[i])
         # Compute the images \phi_T^i for i = 0 .. n.
-        gen_powers = [self(A.gen()**i).coefficients(sparse=False)\
-                        for i in range(0, n + 1)]
-        sys, rhs = Matrix(L, rows, cols), vector(L, rows)
-        rhs[rows - 1] = -1
+        gen_powers = [self(A.gen()**i).coefficients(sparse=False)
+                      for i in range(0, n + 1)]
+        sys, vec = Matrix(K, rows, cols), vector(K, rows)
+        vec[rows - 1] = -1
         for j in range(r):
             for k in range(shifts[j]):
                 for i in range(len(gen_powers[k])):
                     sys[i + n*j, block_shifts[j] + k] = gen_powers[k][i]
         if sys.right_nullity() != 0:
-            raise NotImplementedError(\
-                    "Can't solve system for characteristic polynomial")
-        sol = list(sys.solve_right(rhs))
-        # The system is solved over L, but the coefficients should all lie in Fq
+            raise NotImplementedError("'Gekeler' algorithm failed")
+        sol = list(sys.solve_right(vec))
+        # The system is solved over L, but the coefficients should all be in Fq
         # We project back into Fq here.
-        sol_Fq = list(map(lambda x: L(x).vector()[0], sol))
+        sol_Fq = list(map(lambda x: K(x).vector()[0], sol))
         char_poly = []
         for i in range(r):
-            char_poly.append([sol_Fq[block_shifts[i] + j]\
-                                for j in range(shifts[i])])
+            char_poly.append([sol_Fq[block_shifts[i] + j]
+                              for j in range(shifts[i])])
         return PolynomialRing(self._function_ring, name=var)(char_poly + [1])
-      
+
     def frobenius_norm(self):
         r"""
-        Return Frobenius norm of the Drinfeld module, if the rank is
-        two, raise a NotImplementedError otherwise.
+        Return the Frobenius norm of the Drinfeld module.
 
-        Let `\mathbb{F}_q[T]` be the function ring, write `\chi = X^2 -
-        A(T)X + B(T) \in \mathbb{F}_q[T][X]` for the characteristic
-        polynomial of the Frobenius endomorphism. The *Frobenius norm*
-        is defined as the polynomial `B(T) \in \mathbb{F}_q[T]`.
-
-        Let `n` be the degree of the base field over `\mathbb{F}_q` Then the
-        Frobenius norm has degree `n`.
+        Let `C(X) = \sum_{i=0}^r a_iX^{r-i}` denote the characteristic
```

Comment:
Whare are you using `r-i` instead of `i`?
Also, I would use `is defined as` instead of simply `is`.

Same remark for `frobenius_trace`.

Reply:
Notationally it can actually be a bit cleaner to do it this way, but to ensure consistency with documentation elsewhere I'll revert it.

Reply:
Ok. I don't have a strong opinion on this. At first glance it looks like `i` is a bit easier for the user that won't go into details. That being said, I believe you. @xcaruso, @DavidAyotte: any recommendation?

Reply:
The main gain is that the degree constraint looks a little nicer, and the sum ranges from 1 to r  instead of 0 to r - 1, and reflects the "order" of the term, with the norm being an order r object and the trace being an order 1 object. Not really a huge issue so I'll leave it with the standard convention.